fix: worldport command now teleports players to the expected coordinates

natrist opened the pull request:
<!-- First of all, THANK YOU for your contribution. -->

## Changes Proposed:
This PR proposes changes to:
-  [x] Core (units, players, creatures, game systems).


## Issues Addressed:

1. The worldport command handler's net message reading logic is wrong and the player is not teleported to the requested coordinates
2. Only online players can be teleported

## SOURCE:
<!-- If you can, include a source that can strengthen your claim -->
The changes have been validated through:
- [x] Live research (checked on live servers, e.g Classic WotLK, Retail, etc.)
- [x] Video evidence, knowledge databases or other public sources (e.g forums, Wowhead, etc.)

## Tests Performed:
This PR has been:
- [x] Tested in-game by the author.


## How to Test the Changes:
1. Log into the game world
2. Execute the worldport command (Usage: worldport \<continentID\> [x y z] [facing])


## Known Issues and TODO List:
<!-- Is there anything else left to do after this PR? -->

- [ ] The movecharacter command (NYI) could probably be handled by the worldport command handler later on, as they share the same logic

## How to Test AzerothCore PRs
 
When a PR is ready to be tested, it will be marked as **[WAITING TO BE TESTED]**.

You can help by testing PRs and writing your feedback here on the PR's page on GitHub. Follow the instructions here:

http://www.azerothcore.org/wiki/How-to-test-a-PR

**REMEMBER**: when testing a PR that changes something **generic** (i.e. a part of code that handles more than one specific thing), the tester should not only check that the PR does its job (e.g. fixing spell XXX) but **especially** check that the PR does not cause any regression (i.e. introducing new bugs).

**For example**: if a PR fixes spell X by changing a part of code that handles spells X, Y, and Z, we should not only test X, but **we should test Y and Z as well**.

heyitsbench commented:
Gave this a test. It seems the usage is actually `worldport [MapID] [X] [Y] [Z] [O]`. Providing `603 1897.76 115.94 342.38 0`, for example, took me to Ulduar's Descent into Madness (Vezax's room). Interestingly, whereas `.go xyz` fails to teleport me for not being in a raid group, `worldport` does not fail. Without providing any of those parameters and just the `worldport` command, it took me to the gates of Stormwind regardless of race/class. All looks well. 👍

Curious on the source/use case for `worldport`ing offline players though.

natrist commented:
> Gave this a test. It seems the usage is actually `worldport [MapID] [X] [Y] [Z] [O]`. Providing `603 1897.76 115.94 342.38 0`, for example, took me to Ulduar's Descent into Madness (Vezax's room). Interestingly, whereas `.go xyz` fails to teleport me for not being in a raid group, `worldport` does not fail. Without providing any of those parameters and just the `worldport` command, it took me to the gates of Stormwind regardless of race/class. All looks well. 👍
> 
> 
> 
> Curious on the source/use case for `worldport`ing offline players though.

Yes, that was an oversight on my part. The command indeed takes a mapID as first argument, before the rest of the options. You can also enter a SafeLoc name instead, which is why when you enter nothing it just defaults to the Stormwind GY, as that's probably the first entry in the SafeLoc file.

The worldport command itself does not allow you to teleport other players but the client still sends a null GUID, which is likely because either the GM tool uses the packet for teleporting online players or the handler is also used for the movecharacter command, which does the same thing but on the selected character on the character selection screen.

heyitsbench commented:
> The worldport command itself does not allow you to teleport other players but the client still sends a null GUID, which is likely because either the GM tool uses the packet for teleporting online players or the handler is also used for the movecharacter command, which does the same thing but on the selected character on the character selection screen.

Ah, I misread the second issue the PR addresses in your description then, I apologize. In that case, I have nothing to complain about. 😄

Kitzunu commented on src/server/game/Handlers/MiscHandler.cpp:
```diff
@@ -1117,36 +1117,56 @@ void WorldSession::HandleInspectHonorStatsOpcode(WorldPacket& recv_data)
-    float PositionY;
-    float PositionZ;
-    float Orientation;
+    uint32_t    continentID;
+    uint32_t    requestTimeMs;
+    float       facing;
+    uint64_t    player;
```
remove _t

Kitzunu commented on src/server/game/Handlers/MiscHandler.cpp:
```diff
@@ -1117,36 +1117,56 @@ void WorldSession::HandleInspectHonorStatsOpcode(WorldPacket& recv_data)
+            {
+                uint32_t zoneID = MapMgr::instance()->GetZoneId(0, continentID, position.m_positionX, position.m_positionY, position.m_positionZ);
+
+                CharacterDatabase.Execute(
+                    "UPDATE characters SET map={}, position_x={}, position_y={}, position_z={}, orientation={}, zone={} WHERE guid={}",
+                    continentID, position.m_positionX, position.m_positionY, position.m_positionZ, facing, zoneID, player
```
use TeleportTo() instead

natrist replied:
I'm explicitly porting the player offline. Want me to split the logic into a new function?

natrist commented:
I cleaned up the code and split up the movecharacter command logic into code that I'll submit for another pull request, to keep things tidy.

Here's a video of the movecharacter and worldport commands on WowGM:
https://www.youtube.com/watch?v=r233qne9s6A

Development on the movecharacter command is not yet completed on the client side but the supporting function calls are all implemented.

heyitsbench commented:
No idea if you already know or if this already supports it, but it does seem the original `worldport` command did work even with commas separating the coordinates. E.g. `worldport 1 16223.8, 16259, 13.2022, 88.6758` (macro from older, I believe 3.0, PTR client)

natrist commented:
> No idea if you already know or if this already supports it, but it does seem the original `worldport` command did work even with commas separating the coordinates. E.g. `worldport 1 16223.8, 16259, 13.2022, 88.6758` (macro from older, I believe 3.0, PTR client)

It is not supported but I suspect it could be a side effect of the function converting the string tokens to floating point numbers.

Kitzunu commented on src/server/game/Handlers/MiscHandler.cpp:
```diff
@@ -1117,36 +1117,56 @@ void WorldSession::HandleInspectHonorStatsOpcode(WorldPacket& recv_data)
 }
 
-void WorldSession::HandleWorldTeleportOpcode(WorldPacket& recv_data)
+void WorldSession::HandleWorldTeleport(WorldPacket& message)
```
iirc it is just a naming convension that all opcodes have it in the name

natrist replied:
So you want me to put it back as it was, right?

Kitzunu replied:
ye

Kitzunu replied:
Dont know why recv_data was ignored

natrist replied:
I'm confused. Do you want me to rename the function parameter as well?

Kitzunu commented on src/server/game/Handlers/MiscHandler.cpp:
```diff
@@ -1117,36 +1117,36 @@ void WorldSession::HandleInspectHonorStatsOpcode(WorldPacket& recv_data)
         return;
     }
 
-    LOG_DEBUG("network", "CMSG_WORLD_TELEPORT: Player = {}, Time = {}, map = {}, x = {}, y = {}, z = {}, o = {}", GetPlayer()->GetName(), time, mapid, PositionX, PositionY, PositionZ, Orientation);
```
dont get rid of this, debugs in opcodes are good

natrist replied:
I don't see the point but I don't care, so it's done 😂

Kitzunu commented on src/server/game/Handlers/MiscHandler.cpp:
```diff
@@ -1117,36 +1117,36 @@ void WorldSession::HandleInspectHonorStatsOpcode(WorldPacket& recv_data)
+    if (!playerPtr)
     {
-        LOG_DEBUG("network", "Player '{}' ({}) in flight, ignore worldport command.", GetPlayer()->GetName(), GetPlayer()->GetGUID().ToString());
+        SendNotification("Player not found");
```
```suggestion
        SendNotification("WorldTeleport: Player not found");
```

natrist replied:
I am sending the Player not found notification as that is what the server would normally send on Blizzard realms.

heyitsbench replied:
Is this a debug server-only type string or is this getting sent to players? Because if it's getting sent to players, I feel like there should be some localize-able way to send it.

natrist replied:
Unfortunately it seems like those kinds of things were not localized by Blizzard.

Kitzunu commented:
We can get this merged as soon as you figure out the reviews above...

natrist commented:
> We can get this merged as soon as you figure out the reviews above...

I'm waiting for my other PR that removes redundant network logging calls to get merged. As soon as that's settled I will take another look at the requested changes and reviews.

Kitzunu commented:
Aight, can you ping me in it and I can take a look

heyitsbench commented:
> I'm waiting for my other PR that removes redundant network logging calls to get merged.

Your time has finally arrived.

Nyeriah commented on src/server/game/Handlers/MiscHandler.cpp:
```diff
@@ -1090,34 +1090,34 @@ void WorldSession::HandleInspectHonorStatsOpcode(WorldPacket& recv_data)
+    if (!playerPtr)
     {
-        LOG_DEBUG("network", "Player '{}' ({}) in flight, ignore worldport command.", GetPlayer()->GetName(), GetPlayer()->GetGUID().ToString());
+        SendNotification("Player not found");   //  DO NOT TRANSLATE
```
What's special about this string specifically that it should not be translated?

Kitzunu replied:
Also, you should prefix it with "WORLDPORT: ...."

natrist replied:
In modern internal builds, the client has an extra message handler that displays the console message _Player not found_.
The message does not seem to exist as of the LK alpha/beta builds and no handler can be found for it in patch 3.3.5a.

It is safe to assume the functionality was not present in patch 3.3.5a.
However, we can imagine that Blizzard would still display the error message, as many commands could in theory use it. Add to that the fact that Blizzard sends notifications when cheats are toggled on and off and the theory starts to make sense.

Nyeriah replied:
I appreciate the attention to detail but mimicking their internal builds is not something we aim for when it comes to something that adds no value to the project, like a single error string not localised

Kitzunu replied:
We try to mimic gameplay as much as possible. But admin/gm features etc sometimes can be fun to try and mimic, but errors strings like these are no point in mimicing

sudlud commented:
Closing this for now as it has been abandoned for quite some time.

Please feel free to re-open and update the PR again.
